File: 006/main.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows <= 1:
            return s
        ans = ['' for _ in range(numRows)]
        i, row, direction, sLen = 0, 0, 1, len(s)
        while i < sLen:
            ans[row] += s[i]
            i += 1
            if row == numRows - 1:
                direction = -1
            if row == 0:
                direction = 1
            row += direction
        return ''.join(ans)

if __name__ == '__main__':
    import os
    if os.getenv('LZS'): # local test
        s = Solution()
        assert s.convert('PAYPALISHIRING', 3) == 'PAHNAPLSIIGYIR'
        assert s.convert('PAYPALISHIRING', 1) == 'PAYPALISHIRING'
        assert s.convert('PAYPALISHIRING', 0) == 'PAYPALISHIRING'
        assert s.convert('0123456789', 5) == '0817926354'
        assert s.convert('0123456789', 1) == '0123456789'
        assert s.convert('0123456789', 2) == '0246813579'
        assert s.convert('0123456', 2000) == '0123456'

File: 006/main.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
  if (numRows === 1) {
    return s
  }

  let ans = ''
  let len = 2 * numRows - 2
  let i = 0

  // first row
  while (i < s.length) {
    ans += s[i]
    i += len
  }

  // middle rows
  for (let row = 1; row < numRows - 1; row++) {
    i = row
    while (true) {
      if (i >= s.length) {
        break
      }
      ans += s[i]

      let next = i + (len - 1 - 2 * (row - 1) - 1)
      if (next >= s.length) {
        break
      }
      ans += s[next]
      i += len
    }
  }

  // last row
  i = numRows - 1
  while (i < s.length) {
    ans += s[i]
    i += len
  }
  return ans
}

if (process.env.LZS) { // local test
  const assert = require('chai').assert
  assert.strictEqual(convert('PAYPALISHIRING', 3), 'PAHNAPLSIIGYIR')
  assert.strictEqual(convert('0123456789', 5), '0817926354')
  assert.strictEqual(convert('0123456789', 1), '0123456789')
  assert.strictEqual(convert('0123456789', 2), '0246813579')
  assert.strictEqual(convert('0123456', 2000), '0123456')
}
